enterprise/suppress-user/factory.go

```diff
@@ -21,62 +25,208 @@ type Factory struct {
+		repo,
+		WithLogger(m.Log),
+		WithHttpClient(&http.Client{Timeout: config.GetDuration("HttpClient.suppressUser.timeout", 30, time.Second)}),
+		WithPageSize(config.GetInt("BackendConfig.Regulations.pageSize", 5000)),
+		WithPollIntervalFn(func() time.Duration { return pollInterval }),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return syncer, repo, nil
+}
+
 // Setup initializes the user suppression feature
 func (m *Factory) Setup(ctx context.Context, backendConfig backendconfig.BackendConfig) (types.UserSuppression, error) {
 	if m.Log == nil {
 		m.Log = logger.NewLogger().Child("enterprise").Child("suppress-user")
 	}
+
 	if m.EnterpriseToken == "" {
 		m.Log.Info("Suppress User feature is enterprise only")
 		return &NOOP{}, nil
 	}
+
 	m.Log.Info("Setting up Suppress User Feature")
+
 	backendConfig.WaitForConfig(ctx)
-	var repository Repository
-	if config.GetBool("BackendConfig.Regulations.useBadgerDB", true) {
-		tmpDir, err := misc.CreateTMPDIR()
+
+	var pollInterval time.Duration
+	config.RegisterDurationConfigVariable(300, &pollInterval, true, time.Second, "BackendConfig.Regulations.pollInterval")
+
+	// set up the repository
+	var repo *RepoSwitcher
+	useBadgerDB := config.GetBool("BackendConfig.Regulations.useBadgerDB", true)
+	if useBadgerDB {
+		fullSuppressionPath, latestSuppressionPath, err := getRepoPath()
 		if err != nil {
-			return nil, fmt.Errorf("could not create tmp dir: %w", err)
+			return nil, fmt.Errorf("could not get repo path: %w", err)
 		}
-		path := path.Join(tmpDir, "suppression")
 
-		// TODO: implement seeder source, to retrieve the initial state from a persisted backup
-		var seederSource func() (io.Reader, error)
+		identity := backendConfig.Identity()
 
-		repository, err = NewBadgerRepository(
-			path,
-			m.Log,
-			WithSeederSource(seederSource),
-			WithMaxSeedWait(config.GetDuration("BackendConfig.Regulations.maxSeedWait", 5, time.Second)))
-		if err != nil {
-			return nil, fmt.Errorf("could not create badger repository: %w", err)
+		if _, err = os.Stat(fullSuppressionPath); err != nil && os.IsNotExist(err) {
+			syncer, latestRepo, err := m.NewSyncerWithBadgerRepo(latestSuppressionPath, latestDataSeed, identity, pollInterval)
+			if err != nil {
+				return nil, err
+			}
+
+			subCtx, latestSyncCancel := context.WithCancel(ctx)
+			rruntime.Go(func() {
+				syncer.SyncLoop(subCtx)
+				err = latestRepo.Stop()
+				defer latestSyncCancel()
+			})
+			repo = &RepoSwitcher{
+				Repository: latestRepo,
+				mu:         sync.RWMutex{},
+			}
+			rruntime.Go(func() {
+				syncer, fullRepo, err := m.NewSyncerWithBadgerRepo(fullSuppressionPath, fullDataSeed, identity, pollInterval)
+				if err != nil {
+					m.Log.Error("Complete Synce failed: could not create syncer: %w", err)
+					return
+				}
+				if err = syncer.Sync(ctx); err != nil {
+					m.Log.Error("Complete Synce failed: could not sync: %w", err)
+				}
+				latestSyncCancel()
+				err = repo.Switch(fullRepo)
+				if err != nil {
```

Comment:
is logging enough, or we should handle it any better?

Reply:
Please change the method's signature and stop returning an error. There is no scenario where this could fail

Reply:
damn 🥲